### PR TITLE
fix: Sec. index edge cases in subscriptions and inv. joins

### LINKED ISSUE #4353: Sec. index: Fix all failing cases that are excluded for sec. index multiplier
We added test multiplier for sec. indexes that makes sure all regular (non-index) tests when injected with indexes produce the same result.

Currently some of such tests fail.

Look for `TODO: https://github.com/sourcenetwork/defradb/issues/4353`

### PR DESCRIPTION
**Pull request by islamaliev**
## Relevant issue(s)

Resolves #4353

## Description

Two fixes for cases where the secondary-index test multiplier exposed bugs in the query planner.

**Subscriptions with CID-based queries** were timing out because the planner selected a secondary index even though the `VersionedFetcher` operates on a temporary in-memory store that doesn't contain index data. The index lookup returned empty results, causing the subscription to never fire. The fix skips index selection when CIDs are present, since `VersionedFetcher` only supports the primary index.

**One-to-many inverted joins** returned fewer results because the outer iteration loop and `retrievePrimaryDocs` shared the same plan tree. When `retrievePrimaryDocs` reinitialized the plan, the `orderNode` consumed all fetcher results on its first pass, leaving nothing for subsequent iterations. The fix splits `retrievePrimaryDocs` into two paths: inverted joins use an independent cloned `scanNode` wrapped in an `orderNode`, while non-inverted joins continue using the original plan tree with save/restore. This also eliminates the manual sort that was previously needed — the `orderNode` handles both cases (index-ordered pass-through and in-memory sort).

The upstream fix for scalar filter merging in inverted joins (#4656) is incorporated into the new split-path structure.

Three `MultiplierExcludes` removed.

## How has this been tested?

Existing integration tests pass with and without the secondary-index multiplier. Added deterministic ordering to the one-to-many test to ensure consistent results across index/no-index modes.

Tested on macOS.

### CODE REVIEW

File: tests/integration/txn/patch_collection_test.go [hunk 1]
```diff
@@ -14,16 +14,21 @@ package txn_testing
...
-	"github.com/sourcenetwork/immutable"
 )
 
 // This test runs PatchCollection inside of a transaction, and illustrates that commiting the transaction
 // results in the patch being applied.
 func TestTxn_PatchCollection_WithCommit_Succeeds(t *testing.T) {
 	test := testUtils.TestCase{
+		// The secondary-index multiplier adds @index directives which shifts field indices,
+		// causing the hardcoded "Fields/2" path to target the wrong field.
+		MultiplierExcludes: []string{multiplier.SecondaryIndex},
```
Comment (shahzadlone) [inline tests/integration/txn/patch_collection_test.go]: question: I don't fully understand this, so PatchCollection fails with secondary indexes now?

File: internal/planner/type_join.go [hunk 2]
```diff
@@ -694,21 +706,44 @@ func (r *primaryObjectsRetriever) retrievePrimaryDocs() ([]core.Doc, error) {
 				relFieldName, _ := r.primaryScan.documentMapping.TryToFindNameFromIndex(relFieldIndex)
 				relIDFieldName := request.ToFieldID(relFieldName)
 				relIDFieldMapIndex := r.primaryScan.documentMapping.FirstIndexOfName(relIDFieldName)
-				parentFilter := addFilterOnField(r.filter, r.primarySide.relIDFieldMapIndex.Value(),
-					r.targetSecondaryDoc.GetID())
-				orphan.setSubQueryContext(parentFilter, relIDFieldName, relIDFieldMapIndex)
+				orphan.setSubQueryContext(docFilter, relIDFieldName, relIDFieldMapIndex)
```
Comment (shahzadlone) [inline internal/planner/type_join.go:714]: question: No test hits this line

File: internal/planner/type_join.go [hunk 3]
```diff
@@ -635,16 +630,55 @@ func (r *primaryObjectsRetriever) collectDocs() ([]core.Doc, error) {
...
+) ([]core.Doc, error) {
+	clone := r.primaryScan.cloneWithFilter(docFilter, index, ordering)
+
+	var source planNode = clone
+	if len(ordering) > 0 {
+		source = &orderNode{
+			p:         r.primaryScan.p,
+			plan:      clone,
+			ordering:  ordering,
+			needSort:  true,
```
Comment (fredcarle) [inline internal/planner/type_join.go:664]: question: Shouldn't this be false if the index can satisfy the order? Otherwise there might be an extra in-memory sort that isn't needed.

**Review by fredcarle — APPROVED** [review]
LGTM. Just one question to resolve before merge.

File: tests/integration/txn/patch_collection_test.go (hunk 1 above)
Comment (islamaliev) [reply tests/integration/txn/patch_collection_test.go]: Yeah, not PatchCollection itself, but the test uses a hardcoded JSON pointer path Fields/2 to remove a field. When the secondary-index multiplier adds @index to the SDL, it changes the structure of the Fields array, so index 2 no longer points to the field the test expects. The patch then targets the wrong element and the test breaks. Anything that hardcodes field indices into a patch path is fragile to schema structure changes. Removing the exclude would require rewriting the test to patch by field name instead of index position.

File: internal/planner/type_join.go (hunk 2 above)
Comment (islamaliev) [reply internal/planner/type_join.go:714]: You're right, no test was hitting it in standard coverage runs. It was covered under DEFRA_MULTIPLIERS=secondary-index (via the SUM tests in one_to_many_to_one) but that's not captured by codecov. Added `TestQueryWithNestedOrderByRelationField_ExhaustiveWithPrimaryParentASC_ShouldIncludeOrphans` to cover it explicitly.

File: tests/integration/txn/patch_collection_test.go (hunk 1 above)
Comment (AndrewSisley) [reply tests/integration/txn/patch_collection_test.go]: > When the secondary-index multiplier adds @index to the SDL, it changes the structure of the Fields array, so index 2 no longer points to the field the test expects.

Can you expand on this a little bit please?  How do indexes change the field indexes, and when?  If you add an index after the collection has been created, does it move existing fields?

File: internal/planner/type_join.go (hunk 3 above)
Comment (islamaliev) [reply internal/planner/type_join.go:664]: You're right about the asymmetry. When the index can satisfy ordering, the non-inverted path skips the in-memory sort entirely, but the inverted path was always wrapping in an orderNode regardless.

Your suggested fix of setting `needSort: false` in the constructor wouldn't work though. `Init()` unconditionally resets it to true before iteration starts, so the value gets thrown away before the first `Next()`.

The fix is to skip wrapping in orderNode altogether when canSatisfyOrder is true, mirroring what the non-inverted path does. I've implemented this:                                            

```go
if len(ordering) > 0 && !canSatisfyOrder {
    // only wrap in orderNode when the index can't satisfy ordering
}
```

Thanks for bringing this my attention into it. I had to do some more investigation.

I've also made the ephemeral orderNode iterations visible in the explain output (they were previously invisible since the clone is created at runtime, not in the plan tree). The existing test `TestQueryWithOrderOnOneToMany_WithParentFilterOnRelationAndSubOrder_ShouldOrderChildren` now asserts iterations: 5. Without the optimization it would be 8 (the extra 3 being the redundant orderNode buffering + re-sorting).

File: tests/integration/txn/patch_collection_test.go (hunk 1 above)
Comment (islamaliev) [reply tests/integration/txn/patch_collection_test.go]: hey, guys. When debugging I hastily made incorrect conclusion that the index caused the fields to shift. The more precise reason is that the patching just returns error "field does not exist. Name: name". The reason is that the field is indexes and thus can't be remove. I create an issue to improve error message to be more precise https://github.com/sourcenetwork/defradb/issues/4722 and updated the comments in code
